Source/JavaScriptCore/dfg/DFGMovHintRemovalPhase.cpp

```diff
@@ -59,16 +58,64 @@ class MovHintRemovalPhase : public Phase {
     {
         dataLogIf(DFGMovHintRemovalPhaseInternal::verbose, "Graph before MovHint removal:\n", m_graph);
 
+        // First figure out where various locals are live. We only need to care when we exit.
+        // So,
+        // 1. When exiting, mark all live variables alive. And we completely clear dead variables at that time.
+        // 2. When performing MovHint, it is def and it kills the previous live variable.
+        BlockMap<Operands<bool>> liveAtHead(m_graph);
+        BlockMap<Operands<bool>> liveAtTail(m_graph);
+
+        for (BasicBlock* block : m_graph.blocksInNaturalOrder()) {
+            liveAtHead[block] = Operands<bool>(OperandsLike, block->variablesAtHead, false);
+            liveAtTail[block] = Operands<bool>(OperandsLike, block->variablesAtHead, false);
+        }
+
+        bool changed;
+        do {
+            changed = false;
+            for (BlockIndex blockIndex = m_graph.numBlocks(); blockIndex--;) {
+                BasicBlock* block = m_graph.block(blockIndex);
+                if (!block)
+                    continue;
+
+                Operands<bool> live = liveAtTail[block];
+                for (unsigned nodeIndex = block->size(); nodeIndex--;) {
+                    Node* node = block->at(nodeIndex);
```

Comment:
In general (not for this patch), it would be nice to replace loops like these with iterators. I was very confused about the indexing for a second here, because I missed the last semicolon in the for loop.

Reply:
Maybe we should have some good iterator interfaces for Block etc. (rbegin and rend). Right now this is aligned to the other code.